include/deal.II/matrix_free/cuda_tensor_product_kernels.h

```diff
@@ -90,22 +90,54 @@ namespace CUDAWrappers
        */
       template <int direction, bool dof_to_quad, bool add, bool in_place>
       __device__ void
-      values(Number shape_values[], const Number *in, Number *out) const;
+      values(const Number *in, Number *out) const;
 
       /**
        * Evaluate the gradient of a finite element function at the quadrature
        * points for a given @p direction.
        */
       template <int direction, bool dof_to_quad, bool add, bool in_place>
       __device__ void
-      gradients(Number shape_gradients[], const Number *in, Number *out) const;
+      gradients(const Number *in, Number *out) const;
```

Comment:
So we are reverting the interface that you just introduced?

Reply:
I wanted to save a reference to the right `shape_value` and `shape_gradient` as a class member, but that did not work. Do you know if you can do that for `__constant__` variables?

Reply:
In the case of `MatrixFree` (actual), the`shape_values` and `shape_gradients` are passed by reference in the constructor:
```cpp
    EvaluatorTensorProduct(const AlignedVector<Number2> &shape_values,
                           const AlignedVector<Number2> &shape_gradients,
                           const AlignedVector<Number2> &shape_hessians,
                           const unsigned int            dummy1 = 0,
                           const unsigned int            dummy2 = 0)
      : shape_values(shape_values.begin())
      , shape_gradients(shape_gradients.begin())
      , shape_hessians(shape_hessians.begin())
    {}
```